### PR TITLE
Feat: Bestiary CPP and others

### LINKED ISSUE #1297: Bestiary charm runes.
Charm runes doesnt work on game.

### PR DESCRIPTION
**Pull request by marcosvf132**
First of all, since the bestiary is a system that uses data from all monsters, this PR is uncompleted since i preffer not to include all the monsters changes here for the simple reason that it would be impossible to review it due the size of the changed files (0.5k+ files changed). 

**So you will need to go to this PR #2044 and copy some monsters, so that you can even load you server!!, don't forget to take at least 2 monsters when testing it.**

This PR include:
-Moving the bestiary system from lua to cpp.
-Bestiary charm runes.
-Reflect/heal system by elements.
-Moving the boosted creature from lua to cpp.

If you are using the boosted creature on your website using Lury tutorial, you will need to change few thing on your server:
On the third step, at `boostedMonsterName` you need to replace it by `Game.getBoostedCreature()` and ignore the step 2.

If you find any bug on it, please leave a comment below, if you know any system that uses the old lua bestiary table and i didnt changed it, please tell me so i can fix it.

- Damage reflection monsters lists/percentage comes from this list:
https://tibia.fandom.com/wiki/Damage_Reflection

Resolves #1297

### CODE REVIEW

**Comment by dex-89** [comment]
Can you remove mistakes of Github review? I think about test it.

**Comment by marcosvf132** [comment]
> Can you remove mistakes of Github review? I think about test it.

Which  mistakes? The identation problem? Dont worry, i will fix it, but this do not affect anything when testing it.

**Comment by dex-89** [comment]
> > Can you remove mistakes of Github review? I think about test it.
> 
> Which mistakes? The identation problem? Dont worry, i will fix it, but this do not affect anything when testing it.

Error: [cpplint] reported by reviewdog 🐶
 Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]
```
Raw Output:
src/protocolgame.h:135:  Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]
Error: [cpplint] reported by reviewdog 🐶
 Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]

Raw Output:
src/protocolgame.h:139:  Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]
Error: [cpplint] reported by reviewdog 🐶
 Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]

Raw Output:
src/protocolgame.h:140:  Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]
Error: [cpplint] reported by reviewdog 🐶
 Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]

Raw Output:
src/protocolgame.h:141:  Is this a non-const reference? If so, make const or use a pointer: NetworkMessage &msg  [runtime/references] [2]
Error: [cpplint] reported by reviewdog 🐶
 Add #include <list> for list<>  [build/include_what_you_use] [4]

Raw Output:
src/protocolgame.h:138:  Add #include <list> for list<>  [build/include_what_you_use] [4]
```

**Comment by marcosvf132** [comment]
You can test it even with this reviewdog warning, it wont affect anything in this case.

**Comment by dex-89** [comment]
> You can test it even with this reviewdog warning, it wont affect anything in this case.

So i start tests today 👍

**Review by dex-89 — APPROVED** [review]
19:02 A demon loses 2219 hitpoints due to your attack.
19:02 A demon loses 2219 hitpoints due to your attack.
19:02 A demon loses 2219 hitpoints due to your attack.
19:02 A demon loses 902 hitpoints due to your attack. [Freeze charm]

The charms work, the bestiary works :) so a very good job 👍

> Dont worry, i will fix it

I hold you by the word.

**Comment by gerotib** [comment]
@marcosvf132 Player death is causing server to crash. [Heres crashlog](https://pastebin.com/k19aQVj6). Other than that I see no issues yet.

File: data/scripts/modules/bestiary_charms.lua [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+Bestiaryregister(charm_12, charm_t12)
+--
+-- Bless charm
+--
+local charm_13 = Game.createBestiaryCharm(12)
+local charm_t13 = {}
+
+charm_t13.name = "Bless"
+charm_t13.description = "Triggers on a creature with a certain chance and deals 5% \z
+			of its initial hit points as death damage once."
```
Comment (gerotib) [inline data/scripts/modules/bestiary_charms.lua]:
```suggestion
charm_t13.description = "Blesses you and reduces skill and experience loss by 3% when killed by the chosen creature."
```
Wrong description.

**Comment by gerotib** [comment]
I tested each charm and there are results:

Updated:
wound = works
enflame = works
poison = works
freeze = works
zap = works
curse = works
cripple = works
parry = works
dodge = works
adrenaline = works
numb = works
cleanse = works
bless = works
scavenge = works
gut = works
low blow= works
divine = works
vamp = works
void = works

**Comment by marcosvf132** [comment]
> I tested each charm and there are results:
> 

Thank you for testing it and pointing me the issues.

Fixed all the pointed issues.

File: data/scripts/modules/bestiary_charms.lua [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+--
+-- Cripple charm
+--
+local charm7 = Game.createBestiaryCharm(6)
+local charm_t7 = {}
+
+charm_t7.name = "Cripple"
+charm_t7.description = "Cripples the creature with a certain chance and paralyzes it for 10 seconds."
+charm_t7.type = CHARM_DEFENSIVE
```
Comment (gerotib) [inline data/scripts/modules/bestiary_charms.lua]:
```suggestion
charm_t7.type = CHARM_OFFENSIVE
```
missed change

**Comment by gerotib** [comment]
Another crash in `Combat::doCombatHealth` while trying to attack another player, probably low blow charm issue

```
Thread 2 (Thread 0x7ffff5bae700 (LWP 50911)):
#0  0x00005555557bfe09 in IOBestiary::getCharmFromTarget(Player*, MonsterType*) ()
No symbol table info available.
#1  0x0000555555653a1e in Combat::doCombatHealth(Creature*, Creature*, CombatDamage&, CombatParams const&) ()
No symbol table info available.
#2  0x00005555556e1414 in Player::doAttacking(unsigned int) ()
No symbol table info available.
#3  0x0000555555680a03 in Dispatcher::threadMain() ()
No symbol table info available.
```

**Comment by gerotib** [comment]
Im not sure if thats actual issue, but its not possible to choose charm direclty from bestiary window - [video](https://gfycat.com/beautifulalarmedhammerheadbird). Other than that I see no more issues for now.

**Comment by dex-89** [comment]
> Im not sure if thats actual issue, but its not possible to choose charm direclty from bestiary window - [video](https://gfycat.com/beautifulalarmedhammerheadbird). Other than that I see no more issues for now.

You're right. Charms can be set on fully unlocked monsters [i checked in real Tibia]

**Comment by debone** [comment]
> You can test it even with this reviewdog warning, it wont affect anything in this case.

You can test, but reading the code with varying indentation, multiple patterns of capitilization and variable names that are shorthands or hard to follow makes it so much harder to understand, evaluate and later on change. Can you improve these areas please?

File: data/migrations/7.lua [hunk 3]
```diff
@@ -1,3 +1,30 @@
...
+`adrenaline` INT(250) NULL ,
+`numb` INT(250) NULL, 
+`cleanse` INT(250) NULL ,
+`bless` INT(250) NULL ,
+`scavenge` INT(250) NULL ,
+`gut` INT(250) NULL ,
+`low blow` INT(250) NULL ,
+`divine` INT(250) NULL ,
+`vamp` INT(250) NULL ,
+`void` INT(250) NULL ,
```
Comment (debone) [inline data/migrations/7.lua]: Have you considered creating a table with key values for the charms to avoid having a column for each charm? It looks like the rest of the code follows similar pattern, it may be hard to change, add or a simply misorder break the whole thing.

File: data/scripts/lib/register_bestiary_charm.lua [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  end
+end
+
+registerCharmType.chance = function(charm, mask)
+  if mask.chance then
+    charm:Chance(mask.chance)
+  end
+end
+
+registerCharmType.message_c = function(charm, mask)
```
Comment (debone) [inline data/scripts/lib/register_bestiary_charm.lua]: What do you think about `messageCancel` instead of `message_c`?

File: data/scripts/lib/register_bestiary_charm.lua [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  if mask.message_c then
+    charm:CancelMsg(mask.message_c)
+  end
+end
+
+registerCharmType.message_l = function(charm, mask)
```
Comment (debone) [inline data/scripts/lib/register_bestiary_charm.lua]: Same as cancel message

File: src/combat.cpp [hunk 6]
```diff
@@ -530,6 +534,32 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
...
+		if (player) {
+			if (player->isImmuneCleanse(condition->getType())) {
+				player->sendCancelMessage("You are still immune against this spell.");
+				return;
+			} else if (caster->getMonster()) {
+				MonsterType* mType = g_monsters.getMonsterType(caster->getName());
+				if (mType) {
+					IOBestiary g_bestiary;
+					charmRune_t charm_t = g_bestiary.getCharmFromTarget(player, mType);
+					if (charm_t == CHARM_CLEANSE) {
```
Comment (debone) [inline src/combat.cpp]: I'm thinking of performance of these new checks.

For example here, but the thinking line follows for pretty much all the other charms integrations. 
If you know that you will need `CHARM_CLEANSE` on this line 548, can't your first check on the player binary, up on the line 540, be "Does player has this specific CHARM_CLEANSE active?" and right after "Is the CHARM_CLEANSE monster raceid the target?", both which are relative cheap to get if we take care. Right now `g_bestiary.getCharmFromTarget` while more generic, does some work that is unnecessary.

File: src/enums.h [hunk 7]
```diff
@@ -207,6 +207,66 @@ enum RaceType_t : uint8_t {
...
+	BESTY_RACE_MAGICAL = 14,
+	BESTY_RACE_MAMMAL = 15,
+	BESTY_RACE_PLANT = 16,
+	BESTY_RACE_REPTILE = 17,
+	BESTY_RACE_SLIME = 18,
+	BESTY_RACE_UNDEAD = 19,
+	BESTY_RACE_VERMIN = 20,
+
+	BESTY_RACE_FIRST = BESTY_RACE_AMPHIBIC,
+	BESTY_RACE_LAST = BESTY_RACE_VERMIN,
```
Comment (debone) [inline src/enums.h:267]: What does the word "besty" means?

File: src/IOBestiary.cpp [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+	}
+
+	if (force) {
+		return new Bestiary();
+	}
+
+	return nullptr;
+}
+
+const Bestiary* IOBestiary::getBestiaryCharm(charmRune_t charm_t, bool force /*= false*/) const
```
Comment (debone) [inline src/IOBestiary.cpp]: Duplicated function?

File: src/player.h [hunk 9]
```diff
@@ -1600,6 +1641,108 @@ class Player final : public Creature, public Cylinder
...
+			cleanseCondition.second = OTSYS_TIME() + 10000;
+		}
+		bool isImmuneCleanse(ConditionType_t conditiontype) {
+			uint64_t timenow = OTSYS_TIME();
+			if ((cleanseCondition.first == conditiontype) && (timenow <= cleanseCondition.second)) {
+				return true;
+			}
+			return false;
+		}
+		uint16_t parseRacebyCharm(charmRune_t charmId, bool set, uint16_t newRaceid) {
```
Comment (debone) [inline src/player.h:1688]: Can this function be split into `getRaceByCharm` and `setRaceOnCharm`?

File: src/monsters.h [hunk 10]
```diff
@@ -152,6 +154,18 @@ class MonsterType
 		uint32_t damageImmunities = 0;
 		uint32_t baseSpeed = 200;
 
+		// Bestiary
+		uint8_t Bestiary_Occurrence = 0;
+		uint8_t Bestiary_Stars = 0;
+		uint16_t Bestiary_toKill = 0;
+		uint16_t Bestiary_FirstUnlock = 0;
+		uint16_t Bestiary_SecondUnlock = 0;
```
Comment (debone) [inline src/monsters.h]: I think the standard for variable names is camelCase, can we keep the same pattern? `bestiarySecondUnlock` for example

File: src/IOBestiary.h [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+#include "enums.h"
+#include "monster.h"
+#include "monsters.h"
+#include "game.h"
+#include "player.h"
+#include "luascript.h"
+
+class Game;
+
+class Bestiary
```
Comment (debone) [inline src/IOBestiary.h]: Maybe `BestiaryEntry`? I was confused initially what was this class for.

File: src/combat.cpp [hunk 12]
```diff
@@ -824,6 +854,18 @@ void Combat::doCombatHealth(Creature* caster, Creature* target, CombatDamage& da
 	if(caster && caster->getPlayer()){
 			// Critical damage
 			uint16_t chance = caster->getPlayer()->getSkillLevel(SKILL_CRITICAL_HIT_CHANCE);
+			// Charm low blow rune)
+      if (target && target->getMonster()) {
+			  IOBestiary g_bestiary;
```
Comment (debone) [inline src/combat.cpp]: I'm not sure about memory usage here but do we need to create a variable in every use? IOBestiary is pretty much static, won't we benefit for it to always be around in the many places we are using it pretty much all the time?

File: src/combat.cpp [hunk 13]
```diff
@@ -824,6 +854,18 @@ void Combat::doCombatHealth(Creature* caster, Creature* target, CombatDamage& da
...
+			  MonsterType* mType = g_monsters.getMonsterType(target->getName());
```
Comment (debone) [inline src/combat.cpp]: Have you considered adding `raceId` to the class `Creature`? It seems that we are doing a heavy lift looking for the monster for every time any charm is potentially being affected. That seems impactful.

File: data/migrations/7.lua [hunk 14]
```diff
@@ -1,3 +1,30 @@
 function onUpdateDatabase()
-    return false -- true = There are others migrations file | false = this is the last migration file
+  print("> Updating database to version 8 (Bestiary cpp)")
+  db.query([[CREATE TABLE IF NOT EXISTS `player_charms` (
+`id` INT(250) NOT NULL , 
+`points` VARCHAR(250) NULL , 
+`expansion` BOOLEAN NULL , 
+`wound` INT(250) NULL ,
```
Comment (debone) [inline data/migrations/7.lua]: Maybe if we will keep the runes as columns, prefix the runes with `rune_`? I was looking for why the rune expansion was a boolean... well, it's not a rune :P

File: data/scripts/modules/bestiary_charms.lua [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+local charm_t10 = {}
+
+charm_t10.name = "Adrenaline Burst"
+charm_t10.description = "Bursts of adrenaline enhance your reflexes with a certain chance \z
+			after you get hit and let you move faster for 10 seconds."
+charm_t10.type = CHARM_DEFENSIVE
+charm_t10.chance = 20
+charm_t10.message_c = "Your movements where bursted."
+charm_t10.points = 500
+charm_t10.binary = 512
```
Comment (debone) [inline data/scripts/modules/bestiary_charms.lua]: You can probably deduct the binary flags from the id of the charm. This looks like another place where someone can make a mistake on the order/configuration of charms.

File: src/iologindata.cpp [hunk 16]
```diff
@@ -971,6 +1022,48 @@ bool IOLoginData::savePlayer(Player* player)
...
+  query << "`UnlockedRunesBit` = " << player->UnlockedRunesBit << ',';
+
+  // Bestiary tracker
+  PropWriteStream propBestiaryStream;
+  for (MonsterType* trackedType : player->getBestiaryTrackerList()) {
+   propBestiaryStream.write<uint16_t>(trackedType->info.raceid);
+  }
+  size_t trackerSize;
+  const char* trackerList = propBestiaryStream.getStream(trackerSize);
+  query << "`tracker list` = " << db.escapeBlob(trackerList, trackerSize);
```
Comment (debone) [inline src/iologindata.cpp]: Will it work if player has empty `player->getBestiaryTrackerList()`?

File: src/enums.h (hunk 7 above)
Comment (marcosvf132) [reply src/enums.h:267]: 'Bestiary', i used this abreviation to avoid a conflict with the monsters race (RACE_BLOOD, RACE_VENOM) and to prevent this tag from being real big like 'BESTIARY_RACE_EXTRA_DIMENSIONAL'

File: src/combat.cpp (hunk 13 above)
Comment (marcosvf132) [reply src/combat.cpp]: Well, no. I didn't think about that because this value is only for monsters, so i never considered changing it.

File: src/enums.h (hunk 7 above)
Comment (debone) [reply src/enums.h:267]: I'd lean on the big clear tag :)

File: src/IOBestiary.h [hunk 17]
```diff
@@ -31,13 +31,13 @@
 
 class Game;
 
-class Bestiary
+class Charm
```
Comment (debone) [inline src/IOBestiary.h]: Thanks for this change!

File: src/game.cpp [hunk 18]
```diff
@@ -5738,12 +5744,16 @@ bool Game::combatChangeHealth(Creature* attacker, Creature* target, CombatDamage
...
-				charmRune_t charm_t = g_bestiary.getCharmFromTarget(attackerPlayer, mType);
-				if (charm_t != CHARM_NONE) {
-					Bestiary* charm = g_bestiary.getBestiaryCharm(charm_t);
-					if (charm && charm->type == CHARM_OFFENSIVE && (charm->chance > normal_random(0, 100))) {
-						g_bestiary.ParseCharmCombat(charm, attackerPlayer, target, realDamage);
+			if (!damage.extension && target && target->getMonster()) {
+				MonsterType* mType = g_monsters.getMonsterType(target->getName());
+				if (mType) {
+					IOBestiary g_bestiary;
+					charmRune_t charm_t = g_bestiary.getCharmFromTarget(attackerPlayer, mType);
```
Comment (debone) [inline src/game.cpp]: Using `_t` usually means type. Can it be changed to something like `activeCharm`?

File: src/game.cpp [hunk 19]
```diff
@@ -5738,12 +5744,16 @@ bool Game::combatChangeHealth(Creature* attacker, Creature* target, CombatDamage
...
+					if (charm_t != CHARM_NONE) {
+						Charm* charm = g_bestiary.getBestiaryCharm(charm_t);
+						if (charm && charm->type == CHARM_OFFENSIVE && (charm->chance >= normal_random(0, 100))) {
```
Comment (debone) [inline src/game.cpp:5754]: If I understood this correctly, you can only have one rune on a creature active, right?

File: src/iologindata.cpp [hunk 20]
```diff
@@ -489,29 +489,29 @@ bool IOLoginData::loadPlayer(Player* player, DBResult_ptr result)
...
-	player->charm_bless = result->getNumber<uint16_t>("bless");
-	player->charm_scavenge = result->getNumber<uint16_t>("scavenge");
-	player->charm_gut = result->getNumber<uint16_t>("gut");
-	player->charm_low = result->getNumber<uint16_t>("low blow");
-	player->charm_divine = result->getNumber<uint16_t>("divine");
-	player->charm_vamp = result->getNumber<uint16_t>("vamp");
-	player->charm_void = result->getNumber<uint16_t>("void");
+	player->charm_points = result->getNumber<uint32_t>("charm_points");
+	player->charm_expansion = result->getNumber<bool>("charm_expansion");
+	player->charm_wound = result->getNumber<uint16_t>("rune_wound");
```
Comment (debone) [inline src/iologindata.cpp]: What do you think about changing it on the player class also?  `charm_wound` -> `runeWound`?

File: src/luascript.cpp [hunk 21]
```diff
@@ -14419,6 +14840,178 @@ int LuaScriptInterface::luaMonsterTypeExperience(lua_State* L)
...
+    lua_pushnumber(L, monsterType->info.raceid);
+   } else {
+    monsterType->info.raceid = getNumber<uint16_t>(L, 2);
+    g_game.addBestiaryList(getNumber<uint16_t>(L, 2), monsterType->name);
+    pushBoolean(L, true);
+   }
+  } else {
+   lua_pushnil(L);
+  }
+  return 1;
```
Comment (debone) [inline src/luascript.cpp]: Nitpicking: Indentation here.

File: src/player.h [hunk 22]
```diff
@@ -1794,6 +1939,32 @@ class Player final : public Creature, public Cylinder
 		uint16_t staminaXpBoost = 100;
 		int16_t lastDepotId = -1;
 
+		// Bestiary
+		bool charm_expansion = false;
+		uint16_t charm_wound = 0;
```
Comment (debone) [inline src/player.h]: Extending on my comment that I did on the SQL; Maybe even `charmRuneWound`?

File: src/protocolgame.cpp [hunk 23]
```diff
@@ -559,7 +563,11 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
 		case 0xDC: parseAddVip(msg); break;
 		case 0xDD: parseRemoveVip(msg); break;
 		case 0xDE: parseEditVip(msg); break;
-    case 0xE5: parseCyclopediaCharacterInfo(msg); break;
+		case 0xE5: parseCyclopediaCharacterInfo(msg); break;
+		case 0xe1: BestiarysendRaces(); break;
+		case 0xe2: BestiarysendCreatures(msg); break;
+		case 0xe3: BestiarysendMonsterData(msg); break;
```
Comment (debone) [inline src/protocolgame.cpp]: Nitpicking: These parsers are not camelCase.

File: src/IOBestiary.h [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+class IOBestiary
+{
+	public:
+		Charm* getBestiaryCharm(charmRune_t charm_t, bool force = false);
+		void addBestiaryKill(Player* player, MonsterType* mtype);
+		bool ParseCharmCombat(Charm* charm, Player* player, Creature* target, int32_t realDamage);
+		void addCharmPoints(Player* player, uint16_t amount, bool negative = false);
+		void SendBuyCharmRune(Player* player, charmRune_t runeID, uint8_t action, uint16_t raceid);
```
Comment (debone) [inline src/IOBestiary.h]: Nitpicking: Not all methods are camelCase

File: src/enums.h [hunk 25]
```diff
@@ -207,6 +207,66 @@ enum RaceType_t : uint8_t {
...
+};
+
+enum charm_t {
+	CHARM_UNDEFINED = 0,
+	CHARM_OFFENSIVE = 1,
+	CHARM_DEFENSIVE = 2,
+	CHARM_PASSIVE = 3,
+};
+
+enum races_b : uint16_t {
```
Comment (debone) [inline src/enums.h]: Nitpicking: Enums are usually PascalCase.

File: src/IOBestiary.cpp [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+		charmDamage.extension = true;
+		charmDamage.exString = charm->logMsg;
+
+		charmParams.impactEffect = charm->effect;
+		charmParams.combatType = charmDamage.primary.type;
+		charmParams.aggressive = true;
+
+		player->sendCancelMessage(charm->cancelMsg);
+	} else if (charm->type == CHARM_DEFENSIVE) {
+		switch (charm->id) {
```
Comment (debone) [inline src/IOBestiary.cpp]: I think you could add all the charms under a single switch here and clean the wrapping if else. I was confused at first with the return on line 90 only to understand that it's only about charms that are not OFFENSIVE nor DEFENSIVE. And another reason for the switch is that is not very clear which of the charms are getting to the `doCombatHealth`, which I don't think hurts to be repeated for the sake of clarity.

**Review by debone — COMMENTED** [review]
Thanks for taking my review into account :) I think you did great improvements! I added a few more comments now that we have some other stuff cleared. I'm still very worried about the performance of this one, but for now I run out of ideas.

File: src/combat.cpp (hunk 13 above)
Comment (debone) [reply src/combat.cpp]: Ok, I hear you. Now I noticed another thing, won't `Monster` creatures always have the `->getMonster()` and from there you could get the `MonsterType`? I'm really unsure about using the `getMonsterType` on all these checks.

File: src/game.cpp (hunk 19 above)
Comment (marcosvf132) [reply src/game.cpp:5754]: Yep, you can't have two charm runes into one single monster.

File: src/IOBestiary.cpp (hunk 26 above)
Comment (marcosvf132) [reply src/IOBestiary.cpp]: This is a good point, but if you check the CHARM_OFFENSIVE i dont use switch on it for the simple reason: all of them cause damage except CRIPPLE charm rune, so spliting them i'm avoiding using 7 switch case.

File: src/player.h [hunk 27]
```diff
@@ -343,7 +362,29 @@ class Player final : public Creature, public Cylinder
...
+		}
+
+		void addBestiaryTrackerList(MonsterType* mtype) {
+			auto it = std::find(BestiaryTracker.begin(), BestiaryTracker.end(), mtype);
+			if (it == BestiaryTracker.end()) {
+				BestiaryTracker.push_front(mtype);
+			} else {
+				BestiaryTracker.remove(mtype);
+			}
+			client->refreshBestiaryTracker(BestiaryTracker);
```
Comment (gerotib) [inline src/player.h]:
```suggestion
			if (client) {
				client->refreshBestiaryTracker(BestiaryTracker);
			}
```

File: src/player.h [hunk 28]
```diff
@@ -343,7 +362,29 @@ class Player final : public Creature, public Cylinder
...
+		}
+
+		void sendBestiaryEntryChanged(uint16_t raceid) {
+			client->sendBestiaryEntryChanged(raceid);
```
Comment (gerotib) [inline src/player.h]:
```suggestion
			if (client) {
				client->sendBestiaryEntryChanged(raceid);
			}
```

File: src/player.h [hunk 29]
```diff
@@ -343,7 +362,29 @@ class Player final : public Creature, public Cylinder
...
+		}
+
+		void refreshBestiaryTracker(std::list<MonsterType*> trackerList) {
+			client->refreshBestiaryTracker(trackerList);
```
Comment (gerotib) [inline src/player.h]:
```suggestion
			if (client) {
				client->refreshBestiaryTracker(trackerList);
			}
```

**Review by gerotib — COMMENTED** [review]
I noticed that after crash that happened on startup while game tried to run `payHouses` function that leads to `loadPlayer` and `refreshBestiaryTracker`. Offline player does not have `client` = crash.

**Comment by gerotib** [comment]
Only thing I can think of now is that currently all monster with raceid set can be selected as boosted creature, but according to wiki they are selected from [non-exclusive monsters](https://tibia.fandom.com/wiki/Prey_System/Creature_List#Creatures_that_can_be_rolled_for_free.2Fgold) from prey system. Or at least some flag that could be added to monster file could be useful, because its kind of annoying to have for example butterfly set as boosted creature.

**Review by majestyotbr — APPROVED** [review]

**Comment by majestyotbr** [comment]
Thanks for your contribution!
